### PR TITLE
Remove code duplicates

### PR DESCRIPTION
**Pull request by PatrickHoltz**
# Description

I mainly improved the code of the ObjectManager by removing several code duplicates. To do that I had to create the interface IGameEntity which abstracts objects in the world like npcs, books, teleporters and minigames.

# How Has This Been Tested?

Running the docker-compose-dev.yaml: I visited several books, npcs, minigames in world 1 and its dungeons and compared the behaviour to the data in the lecturer interface.

### CODE REVIEW

File: Assets/Scripts/GameManager/IGameEntity.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+/// <summary>
+/// A GameEntity is an object that is a component of a gameobject placed in the world. Its data is provided by the server.
+/// </summary>
+public interface IGameEntity
+{
+    public void Setup(IGameEntityData data);
```
Comment (delvh) [inline Assets/Scripts/GameManager/IGameEntity.cs:10]: I recommend adding generics `IGameEntity`.
C# has something like `T extends IGameEntityData`, right?
And then, each of the implementations can specify the exact subtype it wants.
The parameter of `Setup` would then simply be `T`.
This has the benefit that you won't need these ugly casts you resort to at the moment.

Comment (PatrickHoltz) [reply Assets/Scripts/GameManager/IGameEntity.cs:10]: Thank you for that hint! I was searching for a way to connect a GameEntity to its GameEntityData but could not find anything.

File: Assets/Scripts/GameManager/Data Classes/WorldData.cs [hunk 2]
```diff
@@ -305,6 +306,42 @@ public DungeonData getDungeonData(int index)
...
+        else
+        {
+            throw new ArgumentOutOfRangeException("There exists no Data Array for " + typeof(T).FullName);
+        }
+        if (index > 0 && index < searchingArray.Length)
+        {
+            return searchingArray[index];
+        }
+        return null;
+    }
```
[the remark this answers is not in the document]
Comment (PatrickHoltz) [reply Assets/Scripts/GameManager/Data Classes/WorldData.cs:343]: I actually created one for the Data classes. It's called IGameEntityData (line 315).
I mean this method exists once for WorldData and for DungeonData but not identically. Where exactly do you see redundancy?

File: Assets/Scripts/GameManager/IGameEntity.cs (hunk 1 above)
Comment (PatrickHoltz) [reply Assets/Scripts/GameManager/IGameEntity.cs:10]: However, this has the disadvantage that you always have to give the IGameEntityData type on every method relating GameEntities even if the data type is not relevant.
Screenshot 2023-01-01 221232

**Review by Gr33ndev — APPROVED** [review]

**Pull request merged**